src/main/java/dev/morling/onebrc/CalculateAverage_elsteveogrande.java

```diff
@@ -0,0 +1,297 @@
+
+            // Read the name's UTF-8 bytes, stop when we hit semicolon
+            int nameSize = 0;
+            while ((b = mmap.get(i++)) != ';') {
+                nameBytes[nameSize++] = b;
+            }
+
+            // Read the temperature reading, stop when we hit the newline
+            int tempSize = 0;
+            while ((b = mmap.get(i++)) != '\n') {
+                tempBytes[tempSize++] = b;
+            }
+
+            // Find (possibly create) this station.
+            var station = getStation(nameBytes, nameSize);
+
+            // Parse temperature reading, and update station's stats
+            var temp = Util.parseFloatDot1(tempBytes);
+            station.update(temp);
+        }
+    }
+
+    Station getStation(byte[] nameBytes, int nameSize) {
+        // (Avoid creating a String unless needed.)
+        return stations.computeIfAbsent(
+                Util.key(nameBytes, nameSize),
+                _ -> new Station(new String(nameBytes, 0, nameSize)));
+    }
+}
+
+final class Util {
+    /**
+     * Convert a string of bytes (but not a {@code String})
+     * into a "key", which is an opaque {@literal long} value.
+     * The rationale is, 64-bit integers as map keys (instead of {@link String}s)
+     * will result in better performance.
+     *
+     * <p>
+     * The returned key is a hash of the input bytes;
+     * therefore it's not guaranteed to be collision-free, but
+     * is "good enough" for our purposes.
+     * </p>
+     */
+    static int key(byte[] bytes, int len) {
+        int i = 0;
+        int ret = 0;
+        byte b;
+        while (i < len) {
+            b = bytes[i++];
+            ret = (59 * ret) + b;
+        }
+        return (int) (ret & 0xffffL);
+    }
+
+    /**
+     * Parse bytes represting a temperature reading into a float.
+     * The input format is expected to be:
+     * <pre><code>
+     *     optionalMinus digit+ literalDot {digit}
+     * </code></pre>
+     * <p>
+     * E.g. {@code -31.4} or {@code 15.9}
+     * </p>
+     */
+    static float parseFloatDot1(byte[] s) {
+        byte b;
+        int i = 0;
+
+        boolean neg = false;
+        if (s[i] == '-') {
+            // Skip the leading minus sign, record that this is negative
+            ++i;
+            neg = true;
+        }
+
+        // Read in base-10 digits, up to the expected '.' char
+        float ret = 0.0f;
+        while ((b = s[i++]) != '.') {
+            ret = (ret * 10.0f) + (float) (b - '0');
+        }
+
+        // Read the final digit which is after the '.'
+        ret += ((s[i] - '0') / 10.0f);
+
+        return neg ? -ret : ret;
+    }
+}
+
+/**
+ * Station names are converted into integers (see {@link Util#key(byte[], int)})
+ * which fit into a known particular range; this "map" will simply index into
+ * an array using these keys to find stations.
+ */
+@SuppressWarnings({ "unchecked" })
+final class ArrayMap<V> extends AbstractMap<Integer, V> {
```

Comment:
Note that this contains an optimization which makes assumptions about the input names.  Station names are hashed and assumed that these uniquely fit within some range (`0` to `65535`), so this wouldn't work in the general case.  If this is "cheating" I can undo that hack :)

Reply:
Changed so that this doesn't rely on the knowledge that there are only ~400 items; can support a more general case now.